### PR TITLE
Add Source::Rewriter APIs for multiple insertions at the same position

### PR DESCRIPTION
**Pull request by yujinakayama**
For https://github.com/whitequark/parser/pull/259#issuecomment-184514530

### CODE REVIEW

**Comment by alexdowad** [comment]
@yujinakayama It's nice that you are working on this. I think the algorithm used is not totally sound, but can't think of a correct one right now.

Another issue: It appears that with this patch, a "multi" insertion will *never* clobber, or be clobbered by, any other edit. That is not correct. A "multi" insertion should not clobber *other insertions*, but (for example) a deletion which spans both sides should still clobber it.

**Comment by yujinakayama** [comment]
> Another issue: It appears that with this patch, a "multi" insertion will never clobber, or be clobbered by, any other edit. That is not correct. A "multi" insertion should not clobber other insertions, but (for example) a deletion which spans both sides should still clobber it.

The multi-insertion properly clobber or be clobbered by other non-insertion actions with the current implementation. I've added specs for those cases.

File: lib/parser/source/rewriter.rb [hunk 1]
```diff
@@ -65,6 +68,28 @@ def insert_before(range, content)
...
 
       ##
+      # Inserts new code before the given source range by allowing other
+      # insertions at the same position.
+      # Note that an insertion with latter invocation comes _before_ earlier
+      # insertion at the same position in the rewritten source.
+      #
+      # @example Inserting '[('
+      #   rewriter.
+      #     insert_after_multi(range, '(').
```
Comment (whitequark) [inline lib/parser/source/rewriter.rb]: `insert_before_multi`

Comment (yujinakayama) [inline lib/parser/source/rewriter.rb]: Fixed

**Comment by yujinakayama** [comment]
Can you cut a release for this please?

**Comment by whitequark** [comment]
@yujinakayama I assume you've tested this with Rubocop?

@JoshCheek Can you please test these changes with your projects and tell if you find them suitable?

**Comment by yujinakayama** [comment]
@whitequark I tested with Transpec, since it does almost the same approach (for analyzing runtime value) with @JoshCheek's SeeingIsBelieving and has issue with the latest parser without the multi-insertion API.

**Comment by whitequark** [comment]
I'd like to get a bit more testing out of this API before committing to support it.

**Comment by JoshCheek** [comment]
I'll try to switch over, but might not have time until next weekend, just super busy right now.

**Comment by JoshCheek** [comment]
Mostly painless ([updates](https://github.com/JoshCheek/seeing_is_believing/commit/4d22850fe75200807bacbb75abe9a4e666ee94fc)). Though note that behaviour changed for inserting before. Not really an issue, just different.

```
[<A>] # old
<[A>] # new
```

**Comment by whitequark** [comment]
You can invoke the lexer alone to extract the comments. That is a supported use case.

**Pull request merged**